cmd/collector/collector.go

```diff
@@ -0,0 +1,178 @@
+	fmt.Fprintf(&buf, "  Sequence No.: %v,  Observation Domain ID: %v\n", msg.SeqNumber, msg.ObsDomainID)
+
+	if msg.Set.GetSetType() == entities.Template {
+		fmt.Fprint(&buf, "TEMPLATE SET:\n")
+		for i, record := range msg.Set.GetRecords() {
+			fmt.Fprintf(&buf, "  TEMPLATE RECORD-%d:\n", i)
+			for _, ie := range record.GetInfoElements() {
+				fmt.Fprintf(&buf, "    %s: len=%d (enterprise ID = %d) \n", ie.Element.Name, ie.Element.Len, ie.Element.EnterpriseId)
+			}
+		}
+	} else {
+		fmt.Fprint(&buf, "DATA SET:\n")
+		for i, record := range msg.Set.GetRecords() {
+			fmt.Fprintf(&buf, "  DATA RECORD-%d:\n", i)
+			for _, ie := range record.GetInfoElements() {
+				fmt.Fprintf(&buf, "    %s: %v \n", ie.Element.Name, ie.Value)
+			}
+		}
+	}
+	klog.Infof(buf.String())
+}
+
+func signalHandler(signalCh chan os.Signal, messageReceived chan *entities.Message) {
+	for {
+		select {
+		case msg := <-messageReceived:
+			printIPFIXMessage(msg)
+		case <-signalCh:
+			break
+		}
+	}
+}
+
+func run() error {
+	klog.Info("Starting IPFIX collector")
+
+	// Load the IPFIX global registry
+	registry.LoadRegistry()
+
+	var netAddr net.Addr
+	var err error
+	if IPFIXTransport == "tcp" {
+		netAddr, err = net.ResolveTCPAddr("tcp", IPFIXAddr+":"+strconv.Itoa(int(IPFIXPort)))
+		if err != nil {
+			return err
+		}
+	} else if IPFIXTransport == "udp" {
+		netAddr, err = net.ResolveUDPAddr("udp", IPFIXAddr+":"+strconv.Itoa(int(IPFIXPort)))
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("input given ipfix.transport flag is not supported or valid")
+	}
+	// Initialize collecting process
+	cp, err := collector.InitCollectingProcess(netAddr, 65535, 0)
+	if err != nil {
+		return err
+	}
+	// Start listening to connections and receiving messages.
+	messageReceived := make(chan *entities.Message)
+	go func() {
+		go cp.Start()
+		msgCount := 0
+		// This can be simplified if collector uses a message channel.
+		// Are there any use cases for having list of messages in collector process.
+		for {
+			ipfixMessages := cp.GetMessages()
+			currentLen := len(ipfixMessages)
+			if currentLen > msgCount {
+				for i := msgCount; i < currentLen; i++ {
+					klog.Info("Processing IPFIX message")
+					messageReceived <- ipfixMessages[i]
+					msgCount++
+				}
+			}
+			time.Sleep(5 * time.Millisecond)
+		}
+	}()
+
+	signalCh := make(chan os.Signal, 1)
+	signal.Notify(signalCh, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
+	go signalHandler(signalCh, messageReceived)
+
+	<-signalCh
+	// Stop the collector process
+	cp.Stop()
+	klog.Info("Stopping IPFIX collector")
+	return nil
+}
+
+func newCollectorCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:  "ipfix-collector",
+		Long: "IPFIX collector to decode the exported flow records",
+		Run: func(cmd *cobra.Command, args []string) {
+			initLoggingToFile(cmd.Flags())
+			if err := run(); err != nil {
+				klog.Fatalf("Error running agent: %v", err)
```

Comment:
Message should be `Error running collector: %v` here

Reply:
Yes :)